rfd/0172-tbot-ssh-muxer.md

```diff
@@ -0,0 +1,171 @@
+connection, it will return after the connection is established. This reduces the
+long-term overhead associated with each connection and the impact on the Linux
+scheduler.
+
+#### Protocol
+
+The Unix socket will implement a simple protocol:
+
+1. Upon connection, the client will send a message indicating the target.
+  a. The message will be encoded in JSON. This will allow more fields to be 
+    added in the future without breaking compatability.
+  b. The `host_port` field will be a string in the format `host:port`.
+2. The server will establish a connection to the target, and then begin
+   forwarding data between the local connection and the target.
+
+Also considered was implementing the SOCKS5 protocol. Whilst this is more 
+standardized, it is more complicated than necessary. It also relies on our
+existing implementation of a SOCKS5 server which has not received signficant
+attention in some time.
+
+#### Client Binary
+
+As it stands today, the `tbot` binary has significant overhead associated with
+it.
+
+Due to this, we will introduce a new, extremely lightweight binary that will
+act as the client.
+
+The client will accept two arguments:
+
+1. The path to the Unix socket.
+2. The message to send to the `tbot` proxy service. This will have the host and
+   port templated into it by OpenSSH with the `%h` and `%p` placeholders.
+
+Comparing Rust and Go for this lightweight binary:
+
+- Rust has a smaller binary size: 400KiB vs 1.8MiB.
+- Rust executes faster.
+- Tuning Go's GC behaviour requires also using `env` as part of the
+  ProxyCommand. This is not necessary with Rust.
+
+For this binary, we will use Rust.
+
+### Initial Scope
+
+For the initial version, we must include:
+
+- The long-lived proxying background service within `tbot`
+- The new lightweight client binary with support for ProxyUseFdPass.
+- A configuration option to control the maximum number of proxied connections.
+  - When this maximum is reached, the service will wait until the number of
+    active connections drops below the maximum.
+  - This will allow a sane limit to be provided to prevent the long-lived `tbot`
+    process from being overwhelmed.
+- Prometheus metrics which exposes the health of the proxying.
+  - Number of active proxied connections.
+  - Number of attempts to establish proxied connections, the time this has
+    taken, and the success status.
+- OpenTelemetry trace spans at key points.
+  - This will allow us to target future performance improvements.
+
+### Future Scope
+
+#### Caching
+
+We can further improve performance by caching the results of operations such as
+resolving the target host from labels.
+
+This work is deferred as it is complex and introduces risks. Information from
+OpenTelemetry tracing will allow us to target the most impactful areas at a 
+later date.
+
+#### Upstream Connection Pooling
+
+The initial version will not share the upstream connections to the Teleport
+Proxy.
```

Comment:
Huh... why?

Reply:
I worry a little bit about performance issues introduced by head of line blocking etc or overuse of the shared connection. If we want to do this, we probably need to carve out some time to determine a safe number of downstream connections to upstream connections. If you're confident we can determine a safe value for this immediately then I'm happy to include this in the work.

Reply:
> Without connection pooling are we really going to reduce resource consumption?

Even without connection pooling we:

- Significantly reduce the work for the Linux scheduler by reducing the number of tasks/processes that are active
- Reducing the amount of memory consumed by the overhead of each of those tbots. Even with the improvements we've made to the heap size, we're probably still using a significant amount more memory than actually necessary to run run a single connection.

@espadolini  is going to run a test with one upstream connection per downstream connection - that'll give us a comparison point and we can make a firmer decision on whether or not the initial iteration needs to include connection sharing.

By "future improvement", I'm not really envisioning this in the far future - more just trying to tighten the scope on the initial version so we don't bite off more than we can chew. Effective connection pooling/sharing seems non-trivial to me - and if there's a significant improvement without adding it, then it seems to me to make more sense for us to work iteratively to keep complexity low. If we can deliver a significant improvement sooner and without complexity, it reduces the pressure on us and means we can spend the time we need ensure that the connection pooling/sharing is stable.

Reply:
Having gone through the figures with Edoardo - I think this is a big enough improvement that we can proceed straight to including a connection cycler.

Reply:
Addressed in [e141c4b](https://github.com/gravitational/teleport/pull/42546/commits/e141c4b06d710f67402f3d49078b7c069643599f)